src/framework/execution/logger.ts

```diff
@@ -120,3 +134,84 @@ function instrumentVerboseTrace(logger: Logger): Logger {
 
   return logger;
 }
+
+function instrumentEventLogging(
+  logger: Logger,
+  eventPublishingQueue: PromiseQueue,
+  inputSynchronizationJobContext?: SynchronizationJobContext,
+): IntegrationLogger {
+  const child = logger.child;
+
+  let synchronizationJobContext:
+    | SynchronizationJobContext
+    | undefined = inputSynchronizationJobContext;
+
+  const publishEvent = (name: string, description: string) => {
+    if (synchronizationJobContext) {
+      const { job, apiClient } = synchronizationJobContext;
+
+      const event = { name, description };
+
+      eventPublishingQueue.add(async () => {
+        try {
+          await apiClient.post(
+            `/persister/synchronization/jobs/${job.id}/events`,
+            {
+              events: [event],
+            },
+          );
+        } catch (err) {
+          // It's not the end of the world if we fail to publish
+          // an event
+          logger.error(
+            {
+              err,
+              event,
+            },
+            'Failed to publish integration event.',
+          );
+        }
+      });
+    }
+  };
+
+  return Object.assign(logger, {
+    flush: async () => {
+      await eventPublishingQueue.onIdle();
+    },
+
+    registerSynchronizationJobContext: (context: SynchronizationJobContext) => {
+      synchronizationJobContext = context;
+    },
+
+    stepStart: (step: IntegrationStep) => {
+      const name = 'step-start';
+      const description = `Starting step "${step.name}"...`;
+      logger.info({ step: step.id }, description);
+
+      publishEvent(name, description);
+    },
+    stepSuccess: (step: IntegrationStep) => {
+      const name = 'step-end';
+      const description = `Completed step "${step.name}".`;
+      logger.info({ step: step.id }, description);
+
+      publishEvent(name, description);
+    },
+    stepFailure: (step: IntegrationStep, err: Error) => {
+      const name = 'step-failure';
+      const description = `Step "${step.name}" failed to complete due to error.`;
```

Comment:
I think we'll soon need an error code that is displayed to users in the event, so that they can provide that to us that we are able to quickly locate the context of the error in the logs. Shall I create a separate Issue to discuss/track that feature?

Reply:
> Shall I create a separate Issue to discuss/track that feature?

Yes please! 

On a related note, I think we should start logging out the integration job id at the start of runs so that users that experience some kind of unexpected issue can also pass that along to us.

Reply:
Great idea, I'll add both. Thanks!